tests: pwm_loopback: Capture disable omission issue

ioannis-karachalios opened the pull request:
1. As per the PWM API definition, `-EBUSY` should be returned when `pwm_enable_capture` is called while capturing
is already enabled. This commit should deal with adding a `ztest_suite_before_t` function that should disable capturing.
Some tests, though, such as `test_pulse_capture`, execute two sub-tests and so, capturing is disabled in between.
In fact, the omission of `pwm_disable_capture` should not only result in aborting a single test, but it can also raise system exception due to invalid context. This is the case for `z_impl_pwm_capture_cycles` where `z_pwm_capture_cycles_callback` can be fired whilst the routine has already aborted (without disabling capturing) and so `struct z_pwm_capture_cb_data data`, defined within function declaration, should not longer be valid.
2. For the inverted pulse measurements, that should reflect the duty-cycle off interval, the 1% deviation in `test_capture` should be adjusted so it reflects duty-cycle off.

Fixes https://github.com/zephyrproject-rtos/zephyr/issues/81199

ioannis-karachalios committed:
tests: pwm_loopback: Capture disable omission issue

As per the PWM API definition, -EBUSY should be returned
when pwm_enable_capture is called and capturing
is already enabled. This commit deals with adding
a ztest_suite_before_t function that should disable
capturing at the end of a single test. Some tests, though,
such as test_pulse_capture, execute two sub-tests
and so, capturing is disabled in between. In fact, the omission
of pwm_disable_capture should not only result in aborting
a single test, but it can also raise system exception
due to invalid context. This is the case for
z_impl_pwm_capture_cycles where z_pwm_capture_cycles_callback
can be fired whilst the routine has already aborted and so
struct z_pwm_capture_cb_data data, defined within function
declaration, should not longer be valid.

Signed-off-by: Ioannis Karachalios <ioannis.karachalios.px@renesas.com>

anangl commented on tests/drivers/pwm/pwm_loopback/src/test_pwm_loopback.c:
```diff
@@ -98,9 +100,15 @@ static void test_capture(uint32_t period, uint32_t pulse, enum test_pwm_unit uni
 	if (flags & PWM_CAPTURE_TYPE_PULSE) {
-		zassert_within(pulse_capture, pulse, pulse / 100,
-			       "pulse capture off by more than 1%");
+		if (flags & PWM_POLARITY_INVERTED) {
+			zassert_within(pulse_capture, period - pulse, (period - pulse) / 100,
+					"pulse capture off by more than 1%");
+		} else {
+			zassert_within(pulse_capture, pulse, pulse / 100,
+					"pulse capture off by more than 1%");
+		}
```
This change is not really related to missing `pwm_disable_capture()` calls. It should be done in a separate commit.

ioannis-karachalios replied:
Moved to a separate commit.

anangl commented on tests/drivers/pwm/pwm_loopback/src/test_pwm_loopback.c:
```diff
@@ -98,9 +100,15 @@ static void test_capture(uint32_t period, uint32_t pulse, enum test_pwm_unit uni
+					"pulse capture off by more than 1%");
+		}
 	}
+
```
And this seems to be unneeded.

ioannis-karachalios replied:
Removed

ioannis-karachalios committed:
tests: pwm_loopback: Fix inverted pulse measurement deviation

For the inverted pulse measurements, that should reflect
the duty-cycle off interval, the 1% deviation in test_capture
should be adjusted so it reflects duty-cycle off.

Signed-off-by: Ioannis Karachalios <ioannis.karachalios.px@renesas.com>

kartben commented:
ping @henrikbrixandersen, if you could please re-approve
